source/scpd/scp/BallotProtocol.d

```diff
@@ -0,0 +1,273 @@
+/*******************************************************************************
+
+    Bindings for scp/BallotProtocol.h
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module scpd.scp.BallotProtocol;
+
+import core.stdc.inttypes;
+
+import scpd.scp.LocalNode;
+import scpd.scp.SCPDriver;
+import scpd.scp.SCP;
+import scpd.scp.Slot;
+
+import scpd.Cpp;
+import scpd.types.Stellar_SCP;
+import scpd.types.Stellar_types;
+import scpd.types.XDRBase;
+
+extern (C++, `stellar`):
+
+// used to filter statements
+//typedef std::function<bool(SCPStatement const& st)> StatementPredicate;
+
+/**
+ * The Slot object is in charge of maintaining the state of the SCP protocol
+ * for a given slot index.
+ *
+ * Defined as a struct to have the right size (`class` always have a vtable)
+ */
+extern(C++, class) public struct BallotProtocol
+{
+    Slot* mSlot;
+
+    bool mHeardFromQuorum;
+
+    // state tracking members
+    enum SCPPhase
+    {
+        SCP_PHASE_PREPARE,
+        SCP_PHASE_CONFIRM,
+        SCP_PHASE_EXTERNALIZE,
+        SCP_PHASE_NUM
+    }
+    // human readable names matching SCPPhase
+    static extern __gshared const(char*)[] phaseNames;
+
+    unique_ptr!SCPBallot mCurrentBallot;      // b
+    unique_ptr!SCPBallot mPrepared;           // p
+    unique_ptr!SCPBallot mPreparedPrime;      // p'
+    unique_ptr!SCPBallot mHighBallot;         // h
+    unique_ptr!SCPBallot mCommit;             // c
+    map!(NodeID, SCPEnvelope) mLatestEnvelopes; // M
+    SCPPhase mPhase;                            // Phi
+    unique_ptr!Value mValueOverride;            // z
+
+    int mCurrentMessageLevel; // number of messages triggered in one run
+
+    /// last envelope generated by this node
+    shared_ptr!SCPEnvelope mLastEnvelope;
+
+    /// last envelope emitted by this node
+    shared_ptr!SCPEnvelope mLastEnvelopeEmit;
+
+  public:
+    //this(Slot& slot);
```

Comment:
Can you add a comment around all the dead code in this PR? Or just remove the code if it will never be needed (if it's needed in the future we should make an issue to revisit this later).